emr-datahub/src/main/scala/org/apache/spark/streaming/aliyun/datahub/DatahubRDD.scala

```diff
@@ -0,0 +1,161 @@
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.streaming.aliyun.datahub
+
+import java.io.UnsupportedEncodingException
+import java.nio.charset.StandardCharsets
+
+import com.aliyun.datahub.DatahubConfiguration
+import com.aliyun.datahub.auth.AliyunAccount
+import com.aliyun.datahub.model.RecordEntry
+import org.I0Itec.zkclient.ZkClient
+import org.I0Itec.zkclient.serialize.ZkSerializer
+import org.apache.spark.internal.Logging
+import org.apache.spark.rdd.RDD
+import org.apache.spark.{InterruptibleIterator, Partition, SparkContext, TaskContext}
+
+import scala.collection.mutable.ArrayBuffer
+
+class DatahubRDD(
+    @transient _sc: SparkContext,
+    duration: Long,
+    endpoint: String,
+    project: String,
+    topic: String,
+    subId: String,
+    accessId: String,
+    accessKey: String,
+    func: RecordEntry => String,
+    shardOffsets: ArrayBuffer[(String, String, String)],
+    zkParam: Map[String, String],
+    checkpointDir: String) extends RDD[Array[Byte]](_sc, Nil) with Logging{
+
+  @transient private var zkClient = DatahubRDD.getClient(zkParam, accessId, accessKey, endpoint)._1
+  @transient private var datahubClientAgent = DatahubRDD.getClient(zkParam, accessId, accessKey, endpoint)._2
+
+  override def compute(split: Partition, context: TaskContext): Iterator[Array[Byte]] = {
+    zkClient = DatahubRDD.getClient(zkParam, accessId, accessKey, endpoint)._1
+    zkClient.setZkSerializer(new ZkSerializer{
+      override def serialize(data: scala.Any): Array[Byte] = {
+        data.asInstanceOf[String].getBytes(StandardCharsets.UTF_8)
+      }
+
+      override def deserialize(bytes: Array[Byte]): AnyRef = {
+        new String(bytes, StandardCharsets.UTF_8)
+      }
+    })
+    datahubClientAgent = DatahubRDD.getClient(zkParam, accessId, accessKey, endpoint)._2
+
+    val partition = split.asInstanceOf[ShardPartition]
+    try {
+      new InterruptibleIterator[Array[Byte]](context, new DatahubIterator(datahubClientAgent, endpoint, project, topic,
+        partition.shardId, partition.cursor, partition.count, subId, accessId, accessKey, func, zkClient,
+        checkpointDir, context))
+    } catch {
+      case e: Exception =>
+        logError("Fail to build DatahubIterator.", e)
+        Iterator.empty.asInstanceOf[Iterator[Array[Byte]]]
+    }
+  }
+
+  override protected def getPartitions: Array[Partition] = {
+    val rate = _sc.getConf.get("spark.streaming.datahub.maxRatePerShard", "10000").toInt
+    var count = rate * duration / 1000
+
```

Comment:
please check if default 10000 is suitable

Reply:
with 60s duration in common,  and 1KB in size, 10000 with make 600MB data in second